sphinx/domains/citation.py

```diff
@@ -0,0 +1,162 @@
+            if data[0] in docnames:
+                self.citations[key] = data
+        for key, data in otherdata['citation_refs'].items():
+            citation_refs = self.citation_refs.setdefault(key, set())
+            for docname in data:
+                if docname in docnames:
+                    citation_refs.add(docname)
+
+    def process_doc(self, env, docname, document):
+        # type: (BuildEnvironment, str, nodes.document) -> None
+        self.process_citations(docname, document)
+        self.process_citation_refs(docname, document)
+
+    def process_citations(self, docname, document):
+        # type: (str, nodes.document) -> None
+        for node in document.traverse(nodes.citation):
+            node['docname'] = docname
+
+            # mark citation labels as not smartquoted
+            label_node = cast(nodes.label, node[0])
+            label_node['support_smartquotes'] = False
+
+            label = label_node.astext()
+            if label in self.citations:
+                path = self.env.doc2path(self.citations[label][0])
+                logger.warning(__('duplicate citation %s, other instance in %s'), label, path,
+                               location=node, type='ref', subtype='citation')
+            self.citations[label] = (docname, node['ids'][0], node.line)
+
+    def process_citation_refs(self, env, docname, document):
+        # type: (BuildEnvironment, str, nodes.document) -> None
+        for node in document.traverse(nodes.citation_reference):
+            target = node.astext()
+            ref = addnodes.pending_xref(target, refdomain='citation', reftype='ref',
+                                        reftarget=target, refwarn=True,
+                                        support_smartquotes=False, ids=node["ids"],
+                                        classes=node.get('classes', []))
+            ref += nodes.inline(target, '[%s]' % target)
+            copy_source_info(node, ref)
+            node.replace_self(ref)
+
+            docnames = self.citation_refs.setdefault(target, set())
+            docnames.add(docname)
+
+    def check_consistency(self):
+        # type: () -> None
+        for name, (docname, labelid, lineno) in self.citations.items():
+            if name not in self.citation_refs:
+                logger.warning(__('Citation [%s] is not referenced.'), name,
+                               type='ref', subtype='citation', location=(docname, lineno))
+
+    def resolve_xref(self, env, fromdocname, builder, typ, target, node, contnode):
+        # type: (BuildEnvironment, str, Builder, str, str, addnodes.pending_xref, nodes.Element) -> nodes.Element  # NOQA
+        docname, labelid, lineno = self.citations.get(target, ('', '', 0))
+        if not docname:
+            if 'ids' in node:
+                # remove ids attribute that annotated at
+                # transforms.CitationReference.apply.
+                del node['ids'][:]
```

Comment:
Before merging, I have to check this is still needed.

Reply:
Now we've been wrap citation labels with `inline` node. As a result, `Element.replace_self()` has no longer warned. So this code is not needed now.